keras_nlp/preprocessing/lm_mask.py

```diff
@@ -0,0 +1,135 @@
+    or used together with `tf.keras.Model` to generate dynamic mask, which is
+    useful for workflows like RoBERTa training.
+
+    Args:
+        vocabulary: A list of strings or a string string filename path. If
+            passing a list, each element of the list should be a single word
+            piece token string. If passing a filename, the file should be a
+            plain text file containing a single word piece token per line.
+        unselectable_tokens: A list of tokens, defaults to None. Tokens in
+            `unselectable_tokens` will not be selected for masking.
+        mask_token: String, defaults to "[MASK]". The mask token.
+    """
+
+    def __init__(
+        self,
+        vocabulary,
+        unselectable_tokens=None,
+        mask_token="[MASK]",
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.unselectable_tokens = unselectable_tokens
+        self.mask_token = mask_token
+
+        # Parse the vocabulary.
+        if isinstance(vocabulary, str):
+            self._vocab = [
+                line.rstrip() for line in tf.io.gfile.GFile(vocabulary)
+            ]
+        elif isinstance(vocabulary, Iterable):
+            self._vocab = list(vocabulary)
+        else:
+            raise ValueError(
+                "Vocabulary must be an file path or list of terms. "
+                f"Received: vocabulary={vocabulary}"
+            )
+
+        # Parse the ids of unselectable tokens.
+        if unselectable_tokens is None:
+            self._unselectable_token_ids = []
+        else:
+            for token in unselectable_tokens:
+                if token not in self._vocab:
+                    logging.warning(
+                        "Unselectable token %s does not exist in the vocab, "
+                        "so it is ignored. Please choose unselectable "
+                        "tokens from the right vocabulary.",
+                        token,
+                    )
+                else:
+                    self._unselectable_token_ids = [
+                        self._vocab.index(token)
+                        for token in unselectable_tokens
+                    ]
+
+        if mask_token not in self._vocab:
+            raise KeyError(
+                f"Mask token {mask_token} does not exist in the given "
+                f"vocabulary, please make sure the mask token is selected "
+                f"from the correct vocabulary."
+            )
+        self._mask_token_id = self._vocab.index(mask_token)
+
+    def call(
+        self,
+        inputs,
+        lm_selection_rate,
+        max_selections,
+        mask_token_rate=0.8,
+        random_token_rate=0.1,
+        padding_token=None,
+    ):
+        if not isinstance(inputs, tf.RaggedTensor):
+            # Convert to RaggedTensor to avoid masking out padded token.
+            inputs = tf.RaggedTensor.from_tensor(inputs, padding=padding_token)
+        random_selector = tf_text.RandomItemSelector(
```

Comment:
Ideally we should support string input and output, which I doubt tf_text supports today. Setting your tokenizer to output strings and doing masking in string space is a really important debugging tool.

This could go one of two ways
 - Add support to tf_text.
 - Redo the logic of these two layers with our own raw tf ops. Which could allow us better string support, dense support and rank 1 support.